Chore: Enable `no-unreduced-motion` and fix errors

ashharrison90 opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

- enables the `no-unreduced-motion` rule in eslint
- fixes _almost_ all the errors
  - there is one in the canvas panel that i think will need the squad to fix themselves
  - have ignored the rule for now, and will raise a followup issue for them to tackle
- updates the lint rule documentation to provide examples of how to fix

**Why do we need this feature?**

- better a11y! 🙌 

**Who is this feature for?**

- everyone!

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

ashharrison90 committed:
enable `no-unreduced-motion` in betterer

ashharrison90 committed:
move all animation calls inside handleReducedMotion

ashharrison90 commented on .betterer.results:
```diff
@@ -826,6 +854,12 @@ exports[`better eslint`] = {
       [0, 0, 0, "\'VerticalGroup\' import from \'@grafana/ui\' is restricted from being used by a pattern. Use Stack component instead.", "0"],
       [0, 0, 0, "\'HorizontalGroup\' import from \'@grafana/ui\' is restricted from being used by a pattern. Use Stack component instead.", "1"]
     ],
+    "packages/grafana-ui/src/components/LoadingBar/LoadingBar.tsx:5381": [
```
the rule looks for all animation properties, i'll shift the others into the `handleReducedMotion`

eledobleefe reviewed (approved):
Looks good to me now 😉

ashharrison90 committed:
fix violations + enable rule

ashharrison90 commented:
@eledobleefe instead of adding it to betterer, i've just fixed the violations and added it to eslint. 

there's 1 that probably needs squad input on how to fix, so i've added a comment, disabled the rule, and will raise a follow up issue for them to look into it.

ashharrison90 committed:
update rule README

eledobleefe commented on packages/grafana-ui/src/components/AutoSaveField/EllipsisAnimated.tsx:
```diff
@@ -22,13 +23,34 @@ const getStyles = () => {
+          animation: `${firstDot} 2s linear infinite`,
+        },
+        {
+          animation: `${firstDot} 2s linear infinite`,
```
I think we can get rid of the animation here since it appears alongside the 'Saving' text 🤔

ashharrison90 replied:
maybe 🤔 

but i'm not sure we need to. everything i've read seems to state that opacity animations are fine. it's more movement/rotation animations that are potentially triggering. that's also what i've updated the rule documentation to state.

eledobleefe replied:
Yes, they are usually related to movement/rotation animations rather than opacity 👌 
I think this answers also my following comment 😄 It's ok to me then, since we can handle them individually whenever we get specific feedback about one of them

eledobleefe commented:
> @eledobleefe instead of adding it to betterer, i've just fixed the violations and added it to eslint.
> 
> there's 1 that probably needs squad input on how to fix, so i've added a comment, disabled the rule, and will raise a follow up issue for them to look into it.

@ashharrison90 I don't want to be a party pooper, but... how do you know which animations to keep when `prefers-reduced-motion` is enabled and which are not? Should they not be disabled by default except for special use cases, such as the "meaningful" animations? 🤔

ashharrison90 commented:
> > @eledobleefe instead of adding it to betterer, i've just fixed the violations and added it to eslint.
> > there's 1 that probably needs squad input on how to fix, so i've added a comment, disabled the rule, and will raise a follow up issue for them to look into it.
> 
> @ashharrison90 I don't want to be a party pooper, but... how do you know which animations to keep when `prefers-reduced-motion` is enabled and which not? Should they not be disabled by default except for special use cases, such as the "meaningful" animations? 🤔

this is the logic i updated the README to use, and is what i went by:

> Animations that involve only non-moving properties, like opacity, color, and blurs, are unlikely to be problematic. In those cases, you still need to wrap the animation in a prefers-reduced-motion media query, but you can use the same animation for both cases:

so any animations that were just toggling opacity/color/etc, i left enabled in both cases. anything that translated/rotated/scaled i either removed entirely or replaced with a more subtle animation. e.g. couple of loading indicators i replaced with a fading in/out animation instead of rotating/bouncing 😂

joshhunt commented on packages/grafana-ui/src/components/Card/CardContainer.tsx:
```diff
@@ -123,9 +133,18 @@ export const getCardContainerStyles = (
-      }),
+      ...handleReducedMotion(
+        {
+          transition: theme.transitions.create(['background-color', 'box-shadow', 'border-color', 'color'], {
+            duration: theme.transitions.duration.short,
+          }),
```
Hmmm. 

I wonder if instead of `handleReducedMotion`  (or as well as) there's something we could add/change to this `theme.transitions.create` thing to handle this. Unsure.

joshhunt commented on packages/grafana-ui/src/components/Collapse/Collapse.tsx:
```diff
@@ -71,7 +72,14 @@ const getStyles = (theme: GrafanaTheme2) => ({
     padding: theme.spacing(1, 2, 1, 2),
     display: 'flex',
-    transition: 'all 0.1s linear',
+    ...handleReducedMotion(
+      {
+        transition: 'all 0.1s linear',
+      },
```
I don't think this transition here even does anything. 

Looking at the code I believe the only thing it transitions is the `display` property, which can't animate anyway.

I think we just get rid of this.

https://github.com/grafana/grafana/blob/7dc6540ead3a03b77308f079b184c5fb5f01020d/packages/grafana-ui/src/components/Collapse/Collapse.tsx/#L70-L79

ashharrison90 commented on packages/grafana-eslint-rules/README.md:
```diff
@@ -26,6 +26,98 @@ To account for users with motion sensitivities, these should always be wrapped i
+// Good ✅
+const getStyles = (theme: GrafanaTheme2) => ({
+  loading: css({
+    ...handleReducedMotion(
+      {
+        animationName: rotate,
+        animationDuration: '2s',
+        animationIterationCount: 'infinite',
+      },
+      {
+        animationName: pulse,
+        animationDuration: '2s',
+        animationIterationCount: 'infinite',
+      }
+    ),
```
@joshhunt yeah so i've been having similar thoughts... 

initially when we were creating this helper function, i was thinking the majority of use cases would have no fallback, i.e. just
```tsx
handleReducedMotion({
  animationName: rotate,
  animationDuration: "2s",
  animationIterationCount: "infinite"
})
```

but actually the majority of cases involve opacity/color transitions that end up with the same styles for both normal and reduced motion. so you end up duplicating code. 

also the more i look at the documentation i added, the vanilla case just using the media queries ends up being clearer and usually has less code which is probably not a good sign 😅 

couple of other options:
- just don't have a helper function at all and rely on the lint rule
  - when added correctly, i find the media query quite readable (do you? 🤔 )
  - i'm just worried people might not click through to the rule documentation where it's explained
- add separate functions to the theme, **very similar to how the breakpoint functions work**
  - this is basically your suggestion, but the important difference would be it wouldn't take the css props as arguments and would just return the string. that way you could use both at once for those cases where the styles should always be applied:
  ```tsx
    // for when both should be applied
    [theme.transitions.prefersMotion, theme.transitions.prefersReducedmotion]: ({
      animationName: pulse,
      animationDuration: "2s",
      animationIterationCount: "infinite",
    }),
    
    // for when separate styles
    [theme.transitions.prefersMotion]: ({
      animationName: rotate,
      animationDuration: "2s",
      animationIterationCount: "infinite",
    }),
    [theme.transitions.prefersReducedMotion]: ({
      animationName: pulse,
      animationDuration: "2s",
      animationIterationCount: "infinite",
    }),
  ```
  - in the end, it's basically just an alias for the media queries, but at least people wouldn't have to google the syntax or refer to the lint rule documentation as much 🤷  

wdyt? 🤔

joshhunt replied:
I like your last suggestion. I think that's the most consistent with what we're doing at the moment, and it leans on the lint rule to catch that it's happening.

Another thing I was considering before was tweaks to `theme.transitions.create()` to accomodate this, but I think your last solution there composes really well with everything else.

joshhunt replied:
Wait - is that valid syntax? 

![image](https://github.com/grafana/grafana/assets/46142/f020b1ca-53b8-4081-959c-0c65fd43ae7e)

😂

ashharrison90 replied:
ah shit, you're right. could make it a function then? something like `[theme.transitions.handleReducedMotion(['no-preference', 'reduced' ])] `

joshhunt replied:
`theme.transitions.handleMotion('no-preference', 'reduced')` ? 

only other thing i can think of is to keep the two static values like you had before, plus a `theme.transitions.motionless`, but i think that's just more akward. 

I think function is better.

joshhunt replied:
ugh. while we're bikeshedding, here's another idea

```ts
css({
  ...theme.transitions.handleMotion('no-preference', {
    transition: '...',
  }),
  ...theme.transitions.handleMotion('motion', {
    transition: '...',
  }),
  ...theme.transitions.handleMotion('motion', 'no-preference',  {
    transition: '...',
  })
})
```

joshhunt replied:
no - just throwing out different options!

gillesdemey commented on public/app/features/alerting/unified/components/rule-editor/alert-rule-form/simplifiedRouting/contactPoint/ContactPointSelector.tsx:
```diff
@@ -172,15 +194,16 @@ const getStyles = (theme: GrafanaTheme2) => ({
-        transform: 'rotate(0deg)',
+    ...handleReducedMotion(
+      {
+        animation: `${rotation} 2s infinite linear`,
```
if possible we prefer to keep the rotation here, if disabled it may look the button isn't doing anything at all. I'm far from an expert on the subject though – do smaller items like this trigger vestibular motion disorder?

ashharrison90 replied:
i think any translation/rotation has the potential to trigger, yes. smaller animations are definitely less bad though

in this case the rotation has been replaced by a fade-in/out pulse, so i think it's unlikely anyone would think it's not doing anything. 

https://www.smashingmagazine.com/2023/11/creating-accessible-ui-animations/#3-essential-animations-that-can-be-adapted is a similar scenario, so i wonder if we can reintroduce the rotation at a slower rate along with the pulse. wdyt? 🤔

ashharrison90 committed:
Merge branch 'main' into ash/reduced-motion-betterer

ashharrison90 committed:
remove unnecessary transition from <Collapse>

ashharrison90 committed:
remove handleReducedMotion utility and add handleMotion to theme

ashharrison90 committed:
update to use new theme value

ashharrison90 committed:
handle Dropdown and IconButton

ashharrison90 committed:
handle AppChromeMenu and update lint message

ashharrison90 committed:
keep rotation at a reduced speed

ashharrison90 commented:
/deploy-to-hg

mckn reviewed (approved):
LGTM! Nice work on this 🙌🏻

ashharrison90 committed:
Merge branch 'main' into ash/reduced-motion-betterer

ashharrison90 committed:
handle DashboardLoading
